### PR TITLE
Fix for windows console 8192 characters limitation stm32 build

### PR DESCRIPTION
**Pull request by JohnieBraaf**
This PR provides a fix for building STM32 ports on Windows using mingw.

The problem:
- Windows Console has a 8192 character limit, which has broken the build process by causing truncation of commands
- The linker command as well as the makemoduledefs commmand can exceed this limit
- See issue: [issue 7687](https://github.com/micropython/micropython/issues/7687) and [issue 7012](https://github.com/micropython/micropython/issues/7012)

The solution
- Storing the collection of files to be passed to linker and makemodulesfs in objs files in the build directory
- See related [Circuit-Python commit](https://github.com/adafruit/circuitpython/commit/7831be5a55baeecb98779fb1c9b62a014f89bdf4)

### CODE REVIEW

**Comment by JohnieBraaf** [comment]
Now made in line with commit guidelines

**Comment by JohnieBraaf** [comment]
The qemu-arm build seems to be failing, even when I revert the changes. Is this a known issue?

make (e=2): The system cannot find the file specified.
make: *** [Makefile:145: run] Error 2

145:  objtuple.o \

**Comment by JohnieBraaf** [comment]
Please ignore the previous comment, I will test qemu-arm on Unix and have a look at the failing qemu-arm/Makefile.test

File: py/py.mk [hunk 1]
```diff
@@ -269,7 +269,10 @@ $(HEADER_BUILD)/compressed.data.h: $(HEADER_BUILD)/compressed.collected
 # build a list of registered modules for py/objmodule.c.
 $(HEADER_BUILD)/moduledefs.h: $(SRC_QSTR) $(QSTR_GLOBAL_DEPENDENCIES) | $(HEADER_BUILD)/mpversion.h
 	@$(ECHO) "GEN $@"
-	$(Q)$(PYTHON) $(PY_SRC)/makemoduledefs.py --vpath="., $(TOP), $(USER_C_MODULES)" $(SRC_QSTR) > $@
+	$(Q)echo $(TOP) > $(BUILD)/top.objs
+	$(Q)echo $(USER_C_MODULES) > $(BUILD)/user_c_modules.objs
+	$(Q)echo $(SRC_QSTR) > $(BUILD)/scr_qstr.objs
+	$(Q)$(PYTHON) $(PY_SRC)/makemoduledefs.py --spath="., @$(BUILD)/top.objs, @$(BUILD)/user_c_modules.objs" @$(BUILD)/scr_qstr.objs > $@
```
Comment (dpgeorge) [inline py/py.mk:275]: `--spath` is not a valid option

also, does it make sense to use `$(ECHO)` instead of `$(Q)echo`?

Comment (JohnieBraaf) [reply py/py.mk]: indeed `$(ECHO)` is more consistent 
and `--spath` was the problem ,both resolved

**Comment by JohnieBraaf** [comment]
I have ran the qemu tests, should be good now, thank you

**Comment by dpgeorge** [comment]
The `unix/extra_coverage.py` test is failing:
```
--- /home/runner/work/micropython/micropython/tests/results/unix_extra_coverage.py.exp	2021-09-16 02:55:29.306564106 +0000
+++ /home/runner/work/micropython/micropython/tests/results/unix_extra_coverage.py.out	2021-09-16 02:55:29.306564106 +0000
@@ -30,14 +30,14 @@
 mport 
 
 builtins        micropython     _thread         _uasyncio
-btree           cexample        cmath           cppexample
-ffi             framebuf        gc              math
-termios         uarray          ubinascii       ucollections
-ucryptolib      uctypes         uerrno          uhashlib
-uheapq          uio             ujson           umachine
-uos             urandom         ure             uselect
-usocket         ussl            ustruct         usys
-utime           utimeq          uwebsocket      uzlib
+btree           cmath           ffi             framebuf
+gc              math            termios         ubinascii
+ucollections    ucryptolib      uctypes         uerrno
+uhashlib        uheapq          uio             ujson
+umachine        uos             urandom         ure
+uselect         usocket         ussl            ustruct
+usys            utime           utimeq          uwebsocket
+uzlib
 ime
 
 utime           utimeq
@@ -159,20 +159,7 @@
 None
 None
 cpp None
-5
-(3, 'hellocpp')
-frzstr1
-frzstr1.py
-frzmpy1
-frzmpy1.py
-frzstr_pkg1.__init__
-frzstr_pkg1/__init__.py 1
-frzmpy_pkg1.__init__
-frzmpy_pkg1/__init__.py 1
-frzstr_pkg2.mod
-1
-frzmpy_pkg2.mod
-1
-ZeroDivisionError
-b'# test frozen package'
-NULL
+Traceback (most recent call last):
+  File "unix/extra_coverage.py", line 53, in <module>
+ImportError: no module named 'cexample'
+CRASH
```
Looks like the USER_C_MODULES are not being included in the build.

**Comment by JohnieBraaf** [comment]
What I was trying will not work for makemodulefs.py. I have reverted this change. 

makemodulefs.py expects a --vpath (comma seperated list of folders) and list of c files as nargs.

It would be possible to write the c files list to a file and pass the file as a parameter to makemodulefs.py.
Would that be an acceptable solution?

**Comment by stinos** [comment]
Just an idea, but the 8192 limit is only for cmd.exe; the win32 process API uses 32768 IIRC (or few less for 0 termination probably) so this makes me wonder: limits for bash running under Cygwin or MSys or so should not be 8192. And 32768 is quite usable already so perhaps another direction would be to figure out which shell can be used to be able to make use of that limit, as it is probably enough for building now?

**Comment by JohnieBraaf** [comment]
@stinos, thank you for the suggestion, I will evaluate if that is a possible solution

**Comment by JohnieBraaf** [comment]
@stinos, you are my hero, MSYS does not exist anymore, but MSYS2 just fantastic.
It has a pacman (Arch like) package manager, and I could just install git/mingw/arm-none-eabi from there.
Most important of all, it does not have any limitation on command size and can use all cores and threads.

As a note, I have tried Cygwin shell, which did not work and only supported a single tread.

I'm closing this PR as to not being needed. 

Should this be noted in the documentation somewhere?

**Comment by stinos** [comment]
:) I really thought the use of MSYS2 was documented somewhere already; at least I've used it extensively. But it's only mentioned in the windows/readme.md and not explicitly for cross-compiling for embedded ports. 

So yes, a PR with a documentation update ('building stm32 port on windows with msys2') for this would be good.

**Pull request closed without merging**